### ISSUE TITLE
Adding Product via GraphQL Stuck in Infinite Loop for Guest Users

### ISSUE THREAD

**Issue by bariyavithal** [Smile-SA/magento2-module-gift-sales-rule#85]
**Pre-conditions:**
Magento version: 2.4.7-p3

**Description**
When attempting to add a product to the cart using GraphQL, the process works correctly for logged-in users, but it gets stuck in an infinite loop for guest users. This behavior occurs when the Smile GiftSalesRule module is enabled and active in Magento 2.

**Steps to Reproduce:**

1. Install and enable the Smile GiftSalesRule module in Magento 2.
2. Create a Cart Price Rule with a gift product condition.
3. Use GraphQL to add a product to the cart:
- For Logged-in Users: Works as expected.
- For Guest Users: The request hangs indefinitely.

**Expected Behavior:**
The product and gift products should be added to the cart successfully for both logged-in and guest users without any delay or infinite loop.

**Actual Behavior:**
For guest users, the process hangs in an infinite loop during the cart update process when the observer associated with `sales_quote_collect_totals_after` is triggered.

**Possible Cause:**
The issue appears to stem from the observer `Smile\GiftSalesRule\Observer\CollectGiftRule`.

**GraphQL Mutation:**

**Create Empty Guest Cart:**

```
mutation {
  createGuestCart {
      cart {
          id
      }
  }
}
```
**Add Gift Product To Guest Cart:**

```
mutation {
  addProductsToCart(
    cartId: "Cy1RIv68dzMXvtk4OB90bbVOBSNPUKup",
    cartItems: [
      {
        quantity: 1
        sku: "PACSHDLF"
      }
    ]
  ) {
    cart {
      itemsV2 {
        items {
          product {
            name
            sku
          }
          quantity
        }
        total_count
        page_info {
          page_size
          current_page
          total_pages
        }
      }
    }
    user_errors {
      code
      message
    }
  }
}
```
Please check the admin rule configuration

![gift-product-rule-configration](https://github.com/user-attachments/assets/562336d8-9329-4efb-8e71-4caa37bab62c)

Would appreciate guidance on how to ensure proper functionality of the CollectGiftRule observer for guest users or if any configuration adjustments are needed.
